### PR TITLE
Accessibility: Adding a `withAssertiveMessages` Higher Order Component

### LINKED ISSUE #2105: Add audible message when links are successfully inserted
Splitting this out from #2098

#2098 is going to make the Links suggestion accessible, replicating the current core implementtion: ARIA combobox + listbox.

screen shot 2017-07-31 at 14 24 45

However, in the current core implementation (see `wp-includes/js/tinymce/plugins/wplink/plugin.js`) a message `Link inserted.` is used when the link is successfully inserted.

I guess this should be done in the other components every time a link is successfully inserted: Enter on the field, click on the button, click/Enter on the highlighted option.

Also, in the current core implementation the highlighted link gets "selected" when pressing the Tab key to move focus to the apply button:

screen shot 2017-07-31 at 14 21 14

I'm not sure if Gutenberg should replicate this behavior, if so it should also use an audible message `Link selected.`

### PR DESCRIPTION
**Pull request by youknowriad**
closes #2105

This PR adds a new Higher Order Component called "withAssertiveMessages" that adds two props "speakAssertive" and "debouncedSpeakAssertive". The pattern of adding these methods was duplicated across multiple components.

This PR tries to resolve #2105 but I'm not sure about the "Term selected" message.

### CODE REVIEW

**Comment by aduth** [comment]
Should this be specific to the assertive tone, or should it be generalized to "speak" (polite and assertive).

**Comment by afercia** [comment]
> or should it be generalized to "speak" (polite and assertive).

From an accessibility perspective, there's the need of both. So far, Gutenberg is mostly using "assertive" to give an immediate audible feedback. In other cases there might be the need to have a "polite" message. The only difference is about timing: `assertive` will stop screen readers from announcing other things they're "speaking" in that moment, while `polite` will make them announce the message only after they've finished announcing other stuff.

**Review by afercia — COMMENTED** [review]
For the a11y part, please see my comment. Fo the architectural part I'd leave that to people more expert than me with React 🙂

**Comment by youknowriad** [comment]
I made the HoC more generic with props taking a second parameter.

File: components/higher-order/with-a11y-messages/index.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+/**
+ * External dependencies
+ */
+import { debounce } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { Component } from 'element';
```
Comment (aduth) [inline components/higher-order/with-a11y-messages/index.js]: Note with #2172, these need to be updated to prefix the dependencies with `@wordpress/`. You will need to perform a rebase against the latest version of master and apply your changes:

```
git fetch origin
git rebase origin/master
```

**Comment by youknowriad** [comment]
If no more concerns, I'll merge this soon.

File: components/higher-order/with-a11y-messages/index.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * @return {Component}                   Component with an instanceId prop.
+ */
+function withA11yMessages( WrappedComponent ) {
+	return class extends Component {
+		constructor() {
+			super( ...arguments );
+			this.debouncedSpeak = debounce( this.speak.bind( this ), 500 );
+		}
+
+		speak( message, type = 'assertive' ) {
```
Comment (aduth) [inline components/higher-order/with-a11y-messages/index.js]: https://github.com/WordPress/gutenberg/pull/2107#issuecomment-319174903 notwithstanding, `packages/a11y` the default type was chosen as `'polite'`

https://github.com/WordPress/packages/blob/2eb0aac/packages/a11y/src/index.js#L31

Inconsistency could be confusing

**Review by aduth — APPROVED** [review]

**Comment by mtias** [comment]
I think with "withSpokenMessage" may be more clear here than just A11y.

**Pull request merged**